### PR TITLE
#164 prototype new inspection: unnecessary else

### PR DESCRIPTION
**Pull request by funivan**
This is just prototype. Need some more work before merge =) 

ToDo
- [x] Improve fixer
- [x] Write documentation
- [x] Check if we should cover alternate `if:else:endif` syntax ?

### CODE REVIEW

File: META-INF/plugin.xml [hunk 1]
```diff
@@ -588,7 +592,10 @@ Bug-fix release, fixed a new bunch of false-positives<br/>
         shortName="DisallowWritingIntoStaticPropertiesInspection" displayName="Disallow writing into static properties"
         groupName="Code Style" enabledByDefault="false" level="WARNING"
         implementationClass="com.kalessil.phpStorm.phpInspectionsEA.inspectors.codeSmell.DisallowWritingIntoStaticPropertiesInspector"/>
-
+    <localInspection language="PHP" groupPath="PHP,Php Inspections (EA Extended)"
```
Comment (kalessil) [inline META-INF/plugin.xml]: This is a duplicate declaration, can be dropped.

File: src/test/resources/fixtures/lang/unnecessary-else.php [hunk 2]
```diff
@@ -0,0 +1,75 @@
+<?php
```
Comment (kalessil) [inline src/test/resources/fixtures/lang/unnecessary-else.php]: This definitely needs time for checking. 

What catches my attention is handling `elseif`: we need an option (disabled by default) to switch on/off `elseif` handling. When it's of if-statement containing `elseif`` should not be inspected at all.

Answering why: code with elseif branching smells as for me, therefore I'd give people illusions that they are improved it by fixing out remarks.

File: src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/languageConstructions/UnnecessaryElseInspector.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiElementVisitor;
+import com.jetbrains.php.lang.lexer.PhpTokenTypes;
+import com.jetbrains.php.lang.psi.elements.*;
+import com.kalessil.phpStorm.phpInspectionsEA.fixers.UnnecessaryElseFixer;
+import com.kalessil.phpStorm.phpInspectionsEA.openApi.BasePhpElementVisitor;
+import com.kalessil.phpStorm.phpInspectionsEA.openApi.BasePhpInspection;
+import org.jetbrains.annotations.NotNull;
+
+/**
```
Comment (kalessil) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/languageConstructions/UnnecessaryElseInspector.java]: Would you please add a header here as it here: https://github.com/kalessil/phpinspectionsea/blob/master/src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/security/CryptographicallySecureAlgorithmsInspector.java#L20-L27
with added `@author ...` of course.

File: src/test/java/com/kalessil/phpStorm/phpInspectionsEA/lang/UnnecessaryElseInspectorTest.java [hunk 4]
```diff
@@ -0,0 +1,16 @@
+package com.kalessil.phpStorm.phpInspectionsEA.lang;
+
+import com.intellij.testFramework.fixtures.CodeInsightFixtureTestCase;
+import com.kalessil.phpStorm.phpInspectionsEA.inspectors.languageConstructions.UnnecessaryElseInspector;
+
+/**
+ * (c) Funivan <alotofall@gmail.com>
+ */
+public class UnnecessaryElseInspectorTest extends CodeInsightFixtureTestCase {
```
Comment (kalessil) [inline src/test/java/com/kalessil/phpStorm/phpInspectionsEA/lang/UnnecessaryElseInspectorTest.java]: `final` + I dropping all class comments in tests normally. There is git history to track an author.

File: src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/languageConstructions/UnnecessaryElseInspector.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+                visitElseKeywords(elseStatement);
+            }
+
+            @Override
+            public void visitPhpElseIf(ElseIf elseIfStatement) {
+                super.visitPhpElseIf(elseIfStatement);
+                visitElseKeywords(elseIfStatement);
+            }
+
+            private void visitElseKeywords(Statement elseStatement) {
```
Comment (kalessil) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/languageConstructions/UnnecessaryElseInspector.java]: Please check this: https://github.com/kalessil/phpinspectionsea/blob/master/src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/ifs/IfReturnReturnSimplificationInspector.java#L37-L40

General agreement (would you please mention this in dev docs?) that we are not analyzing control structures without `{}`.

Not for documenting:  we have an inspection complaining about missing `{}`, if someone ignores it, we should not support him/her. And ofcourse the logic/performance will be pretty slim with this approach.

File: src/test/resources/fixtures/lang/unnecessary-else.php (hunk 2 above)
Comment (funivan) [inline src/test/resources/fixtures/lang/unnecessary-else.php]: Done

File: src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/languageConstructions/UnnecessaryElseInspector.java (hunk 5 above)
Comment (funivan) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/languageConstructions/UnnecessaryElseInspector.java]: I still have method `hasBraces` to check if we try to handle code without braces.  Still do not know how to check this with `ExpressionSemanticUtil.getGroupStatement`

Comment (kalessil) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/languageConstructions/UnnecessaryElseInspector.java]: Will provide an example today, ping me tomorrow if i forgot please

Comment (funivan) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/languageConstructions/UnnecessaryElseInspector.java]: ok =)

Comment (funivan) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/languageConstructions/UnnecessaryElseInspector.java]: @kalessil ping

Comment (kalessil) [inline src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/languageConstructions/UnnecessaryElseInspector.java]: Drawn a big cross on my hand %)

**Comment by kalessil** [comment]
So we can start now: please target the branch feature-164 here and I'll make a PR with code samples I mentioned before

**Comment by funivan** [comment]
Target branch was changed

**Comment by kalessil** [comment]
Merged, will make a PR into this branch and we'll step by step review it (at Skype probably will be best way, but let me make a PR first).

**Pull request merged**